Fix WinMD handling in System.Reflection.MetadataLoadContext

andrew-boyarshin opened the pull request:
Fixes #58319

Add a test WinMD file to prevent regressions. Had to be compiled with .NET Framework ILASM, since .NET ILASM silently ignores `/MDV` switch, which is necessary to produce WinMD that reproduces the crash (`MetadataVersion` is used to determine the existence of _virtual AssemblyRefs_).

andrew-boyarshin committed:
Fix WinMD handling in System.Reflection.MetadataLoadContext

Add a test to prevent regressions

am11 commented on src/libraries/System.Reflection.MetadataLoadContext/tests/src/Tests/Type/TypeTests.cs:
```diff
@@ -611,5 +616,30 @@ public static void TypesWithStrangeCharacters()
+            PathAssemblyResolver resolver = new(runtimeAssemblies);
+            using MetadataLoadContext mlc = new(resolver);
+
+            Stream peStream = new MemoryStream(TestData.s_MetadataDelegateImage);
```
nit:
```suggestion
            using MemoryStream peStream = new(TestData.s_MetadataDelegateImage);
```

andrew-boyarshin committed:
Apply review suggestion

andrew-boyarshin commented:
`runtime (Libraries Build Linux x64 Debug)` failure appears to be #55449.

andrew-boyarshin committed:
Skip the new test on Browser platform

The MetadataLoadContext should work on Browser, but the test requires
better MetadataAssemblyResolver implementation

andrew-boyarshin commented:
Can this make it into 7.0.0 milestone? It's a pretty simple fix for a bug that makes it impossible to properly process any Windows Runtime WinMD file using `System.Reflection.MetadataLoadContext`.

steveharter commented:
@andrew-boyarshin can you explain a bit your scenario? I assume some UWP work given the original `Windows.Foundation.winmd` file specified in the issue. Thanks.

andrew-boyarshin committed:
Address review feedback

andrew-boyarshin committed:
Bring back the test skip for Browser

System.Runtime still fails to resolve on Browser.
WinMD parsing on Browser is questionable enough. Though, with better
MetadataAssemblyResolver impl it should still work. Making this new
unit test be Browser-aware is out-of-scope for this fix.
The issue to be fixed is not related to assembly loading at all.

AaronRobinsonMSFT commented:
I've been told the WinMD format has enough differences with ECMA-335 that support should be considered a non-goal. Not saying we shouldn't accept this but does this actually represent full support?

/cc @jkotas @jkoritzinsky

jkotas commented:
How would you explain (e.g. in documentation) what is and what is not supported after these fixes?

The most complex part of the WinRT interop support are type projections. This does not seem to be handling the type projections at all. I doubt that you can actually use this as .NET type system without the type projections.

We have intentionally removed the built-in support for WinRT interop in https://github.com/dotnet/runtime/issues/35318 . The WinRT interop is being developed independently on .NET runtime in https://github.com/microsoft/CsWinRT/ .

andrew-boyarshin commented:
@jkotas this is independent of WinRT interop support removed in 5.0. This is merely a library that leverages `System.Reflection` to load .NET-like PE assemblies and builds the type system for reflection. `System.Reflection` should already support everything needed for that. In fact, not using the helpers from `System.Reflection` was the cause of the issue fixed in this PR.
`System.Reflection.MetadataLoadContext` doesn't need type projections, since it's not an interop solution, but an assembly inspection library that provides building blocks for something like ildasm. ildasm doesn't need type projections, because it should display what is actually in the assembly. WinMD format was deliberately chosen to be a preexisting format to leverage the existing tooling. Unfortunately, a few extensions to .NET PE format were made, such as _virtual_ AssemblyRefs, which is not a problem for `System.Reflection`, but before this PR, was a problem for `System.Reflection.MetadataLoadContext`. This is not a feature PR, but a bugfix PR to enable users of the old .NET Framework reflection-only assembly loading contexts to migrate to modern .NET. You can build an interop solution (do AOT codegen based on the type system provided by `System.Reflection.MetadataLoadContext`), but it's not something `System.Reflection.MetadataLoadContext` provides out-of-box, nor should it provide that, in my opinion.

jkotas commented:
> I have presented a patch that allowed me to parse an entire Windows Runtime API surface with a high-level API. 

The point of this discussion is whether we are comfortable with supporting this scenario going forward. WinRT is developed independently on .NET these days.  System.Reflection.MetadataLoadContext was not designed for reading .winmd files. There are scenarios today where System.Reflection.MetadataLoadContext won't work well for reading .winmd files, and this set is likely to grow over time.

> It also reduces code complexity by reusing existing code and dropping generic parameter 

This simplification is definitely welcomed. Thank you!

>  I'd be interested to read about the differences between .NET type system and WinRT type system

I am not aware of comprehensive up-to-date document that lists all differences. I can give you one example: Names in assembly references are irrelevant in WinRT type system. The name in .winmd assembly references maps to correct file name some of the time. It is fragile to depend on it. The WinRT type system specifies [algorithm](https://docs.microsoft.com/en-us/uwp/winrt-cref/winmd-files#winmd-file-name) for locating .winmd files. The runtime built-in WinRT interop support had methods like `AppDomain.TypeResolve` to support the WinRT specific algorithm for locating .winmd files. None of this support is present in System.Reflection.MetadataLoadContext.

Also, I was wrong that `System.Reflection.MetadataLoadContext` does not apply type projections. It does apply the type projections since `MetadataReaderOptions.Default = MetadataReaderOptions.ApplyWindowsRuntimeProjections`. However, this is the frozen snapshot of the type projections that matches the runtime built-in WinRT interop support. The set of WinRT type projections was extended since then, and so this won't work well for more recent WinRT types such like the ones that come with WinUI 3.

steveharter commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/Ecma/MetadataTable.cs:
```diff
@@ -26,12 +25,12 @@ public MetadataTable(int count)
             Debug.Assert(!handle.IsNil);
             Debug.Assert(factory != null);
 
-            int index = handle.GetToken().GetTokenRowNumber() - 1;
```
> System.Reflection handles WinMDs fine, because it knows about virtual AssemblyRefs

I assume you mean `System.Reflection.Metadata`, not `System.Reflection` via the reader option `MetadataReaderOptions.ApplyWindowsRuntimeProjections` and the resulting `MetadataKind.WindowsMetadata` enum value.

steveharter commented:
Attempt at summary:
- Workaround is to use `MetadataReader` and apply WinMD semantics appropriately.
- WinRT runtime interop removed in #35318, although this PR is about the raw metadata format not runtime. WinRT interop is being developed independently on .NET runtime in https://github.com/microsoft/CsWinRT/. Also see also see see https://docs.microsoft.com/en-us/dotnet/core/compatibility/interop/5.0/built-in-support-for-winrt-removed
- `MetadataReader` has some support for windows metadata via `MetadataReaderOptions.ApplyWindowsRuntimeProjections` but those for earlier point-in-time interop scenarios. MLC uses that option.
- There are other cases where WinMD files won't work including basic Type names including base types (those are supposed to map to .WinMD files) and with additional to the more recent WinUI 3. Due to this, even with this PR, the resulting reflection types may not match to the expected WinMD reflection types without properly supporting type projections.

buyaa-n commented:
@andrew-boyarshin  with the above reasons mentioned in the @steveharter's `Attempt at summary` we are hesitant to accept this fix, as we did not receive any response from you for a while added `no recent activity` label so that the PR could get closed automatically if there is no activity within a week

buyaa-n commented:
Closing, somehow the automation is not working for this PR
